contracts/src/Gateway.sol

```diff
@@ -412,13 +412,14 @@ contract Gateway is IGateway, IInitializable {
     }
 
     // @dev Register a new fungible Polkadot token for an agent
-    function setTokenOwnerAgentID(address token, bytes32 agentID) external {
+    function registerToken(bytes32 tokenID, address token, bytes32 agentID) external {
```

Comment:
I've thought of a way we can prevent malicious parachains from using `Transact` to call back into gateway via an agent.

In AgentExecutor.sol, we make sure the destination of the transact is not the gateway contract. If it the destination is the gateway, we disallow the transact.

What we still need for this `registerToken` is a guard to ensure it's only being called by a registered agent. We may need to introduce another storage field which allows to perform this lookup. Unfortunately the storage field below only allows us to lookup by agentID:

https://github.com/Snowfork/snowbridge/blob/8b8f1b5af3fc0a622f207d8069a04c0821ad67ce/contracts/src/storage/CoreStorage.sol#L16

Reply:
We receive the relevant `agentID` as the 3rd parameter here, so we can check that `msg.sender == agents[agentID]`. This means that only a registered agent can call `registerToken` and that they must register tokens for themselves. It's slightly stricter than just "registered agents only", but seems reasonable as the intention is for agents to register their own tokens anyway.

Reply:
> prevent malicious parachains from using Transact to call back into gateway via an agent. In AgentExecutor.sol, we make sure the destination of the transact is not the gateway contract. If it the destination is the gateway, we disallow the transact.

@vgeddes Don't think we can prevent arbitrary transact from reentering Gateway without a [ReentrencyGuard](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/security/ReentrancyGuard.sol),  demonstrated in https://github.com/Snowfork/snowbridge/pull/933